src/topology/homotopy/basic.lean

```diff
@@ -0,0 +1,303 @@
+
+lemma prop (F : homotopy_with f₀ f₁ P) (t : I) : P (F.curry t) := F.prop' t
+
+@[simp]
+lemma curry_apply (F : homotopy_with f₀ f₁ P) (t : I) (x : X) : F.curry t x = F (t, x) := rfl
+
+/--
+Continuously extending a curried homotopy to a function from `ℝ` to `C(X, Y)`.
+-/
+def extend (F : homotopy_with f₀ f₁ P) : C(ℝ, C(X, Y)) := F.curry.Icc_extend zero_le_one
+
+lemma extend_apply_of_le_zero (F : homotopy_with f₀ f₁ P) {t : ℝ} (ht : t ≤ 0) (x : X) :
+  F.extend t x = f₀ x :=
+begin
+  rw [←F.apply_zero],
+  exact continuous_map.congr_fun (set.Icc_extend_of_le_left (@zero_le_one ℝ _) F.curry ht) x,
+end
+
+lemma extend_apply_of_one_le (F : homotopy_with f₀ f₁ P) {t : ℝ} (ht : 1 ≤ t) (x : X) :
+  F.extend t x = f₁ x :=
+begin
+  rw [←F.apply_one],
+  exact continuous_map.congr_fun (set.Icc_extend_of_right_le (@zero_le_one ℝ _) F.curry ht) x,
+end
+
+@[simp]
+lemma extend_apply_coe (F : homotopy_with f₀ f₁ P) (t : I) (x : X) : F.extend t x = F (t, x) :=
+continuous_map.congr_fun (set.Icc_extend_coe (@zero_le_one ℝ _) F.curry t) x
+
+@[simp]
+lemma extend_apply_of_mem_I (F : homotopy_with f₀ f₁ P) {t : ℝ} (ht : t ∈ I) (x : X) :
+  F.extend t x = F (⟨t, ht⟩, x) :=
+continuous_map.congr_fun (set.Icc_extend_of_mem (@zero_le_one ℝ _) F.curry ht) x
+
+lemma extend_prop (F : homotopy_with f₀ f₁ P) (t : ℝ) : P (F.extend t) :=
+begin
+  by_cases ht₀ : 0 ≤ t,
+  { by_cases ht₁ : t ≤ 1,
+    { convert F.prop ⟨t, ht₀, ht₁⟩,
+      ext,
+      rw [F.extend_apply_of_mem_I ⟨ht₀, ht₁⟩, F.curry_apply] },
+    { convert F.prop 1,
+      ext,
+      rw [F.extend_apply_of_one_le (le_of_not_le ht₁), F.curry_apply, F.apply_one] } },
+  { convert F.prop 0,
+    ext,
+    rw [F.extend_apply_of_le_zero (le_of_not_le ht₀), F.curry_apply, F.apply_zero] }
+end
+
+end
+
+variable {P : C(X, Y) → Prop}
+
+/--
+Given a continuous function `f` and a proof `hP : P f`, we can define a `homotopy_with f f` by
+`F (t, x) = f x`
+-/
+def refl (f : C(X, Y)) (hP : P f) : homotopy_with f f P :=
+{ to_fun := λ x, f x.2,
+  continuous_to_fun := by continuity,
+  to_fun_zero := λ _, rfl,
+  to_fun_one := λ _, rfl,
+  prop' := λ t, begin
+    convert hP,
+    ext, refl,
+  end }
+
+instance : inhabited (homotopy_with (continuous_map.id : C(X, X)) continuous_map.id (λ f, true)) :=
+  ⟨homotopy_with.refl continuous_map.id trivial⟩
+
+/--
+Given a `homotopy_with f₀ f₁ P`, we can define a `homotopy_with f₁ f₀ P` by reversing the homotopy.
+-/
+def symm {f₀ f₁ : C(X, Y)} (F : homotopy_with f₀ f₁ P) : homotopy_with f₁ f₀ P :=
+{ to_fun := λ x, F (σ x.1, x.2),
+  continuous_to_fun := by continuity,
+  to_fun_zero := by norm_num,
+  to_fun_one := by norm_num,
+  prop' := λ t, by simpa using F.prop' (σ t) }
+
+@[simp]
+lemma symm_apply {f₀ f₁ : C(X, Y)} (F : homotopy_with f₀ f₁ P) (x : X) (t : I) :
+  F.symm (t, x) = F (σ t, x) := rfl
+
+@[simp]
+lemma symm_symm {f₀ f₁ : C(X, Y)} (F : homotopy_with f₀ f₁ P) : F.symm.symm = F :=
+begin
+  ext x,
+  cases x,
+  simp,
+end
```

Comment:
```suggestion
by { ext ⟨t, x⟩, simp }
```

Reply:
actually I think here it makes sense to change the `simp` lemma instead.